### ISSUE TITLE
Lambda function coldstart taking too long

### ISSUE THREAD

**Issue by leothomas** [NASA-IMPACT/nasa-apt#314]
Apparently cold starts are known to be worst with docker image lambdas: https://mikhail.io/serverless/coldstarts/aws/
^ according to this article, in the case of docker, it doesn't really seem to be related to deployment package size

Some possible solutions: 
- Implement CloudFront cache
- Use cloudwatch/event bridge events to ping lambda every 5-10 minutes: https://epsagon.com/development/how-to-minimize-aws-lambda-cold-starts/

**Comment by aboydnw** [comment]
@leothomas I pulled this one into scope for the Pre-Launch list of items because I think it relates to Option 2 that we talked about last week to address some of the instability in APT. 
https://docs.google.com/document/d/1WOoS02qCWIs-ne6BeQNuA2gV00nVZnXTpAjrUf3MLLY/edit?usp=sharing

I think you used the phrase "splitting the lambdas" although I am not entirely sure what that means or entails. Can you confirm if this is a good ticket to represent that work? If not, can you make one?

Also let me know if you don't recall that conversation or if I am not making any sense.

**Comment by leothomas** [comment]
I definitely recall that conversation, and you are definitely making sense!

This ticket is more so tracking the application coldstarts: sometimes, when _no one_ has used the application in a while, the API will inexplicably take 10-15 seconds to respond. Re-reading the initial ticket I put together, I'm realizing that there is an added layer of complexity: I thought it was obvious that these 10-15 seconds delays were due to the Lambda function coldstarts (when a lambda function instantiates a new container to run in), however looking into the logs, the Lambda function coldstarts are never more than 3-4 seconds, so the 10-15 seconds delay is coming from somewhere else. 

The 3-4 second Lambda coldstarts are likely made worse by the Lambda function's package size (huge because of all the Latex related libraries), and that splitting up the Lambdas such that all PDF generation code (and Latex libraries) are in a separate lambda function would mean that that 3-4 second cold start would only ever occur when generating a PDF (which is much less noticeable, since the PDF generation already takes a little while).

While splitting up the lambda guarantees to resolve the lambda function's coldstarts (3-4 seconds), it doesn't _guarantee_ to resolve the problem with the application sometimes taking 10-15 seconds to load, although I suspect that it might. 

With that being said, when I brought up API stability, I was referring to several different little things including: 
- Sometimes PDF generation fails with an internal server error, but if you refresh it generates immediately
- Sometimes PDF generation fails when data is missing, instead of just generating a document with a blank section

Splitting up the lambdas _might_ help with these, but is by no means guaranteed to.

So to summarize: 

This ticket more so tracks random 10 second delays when no has use the app in a little while - this can be improved (and _maybe_ resolved) by splitting up the lambda functions. 

Splitting up the lambda functions _might_ help with random errors when trying to generate PDFs. 

I would consider "random errors when trying to generate PDFs" to be better represented by [this ticket](https://github.com/NASA-IMPACT/nasa-apt/issues/471)

**Comment by jo-tham** [comment]
I know this issue is on svalbard. Not much on backend for byblos so I thought I'd check some more.

If you could use more assistance here, I have a couple questions/comments

1. can you post some aws cloudwatch metrics on response time/latency to provide more context here? it would just be interesting to know how often this happens. once it's in production or in use more, is it possible this won't really be a problem? Is there a possible frontend solution, like after 3 seconds of waiting, adding an additional message to the effect of 'sorry, we're cold starting'
   a. ideally these metrics would be segmented by route (is it just specfic routes that have this long delay?)

2. If you want to try the lambda zip route and including latex, i might be able to help. i've done that with tensorflow-lite, i.e. run a local container that emulates (or is) the aws lambda runtime and build the dependency binaries. But @aboydnw should know that is a potentially large effort uncertainty - could take a few days to iron out the issues and still find out it's too large for the zip package


### tangential architecture remarks 

Ineed @leothomas i think using api gateway to define the routes and splitting the lambads up to minimize dependencies for each resource (e..g only the pdf lambda provides latex in its runtime) would be beneficial here and a good refactor general. It comes with additional benefit of letting aws handle more of the api boilerplate, including
  - cognito auth & resource access control (we probably wouldn't have to handle this in the application code)
  - rate limiting (cost management measure, for when it's in the wild, bad actors, and cost control)
  - kind of minimize the application logic/error surface

Clearly 2. is beyond the scope of a solution to this problem. just wanted to talk about it a bit further. 

I see in the readme that we used to use persistent servers/ecs, so routing framework wasn't part of the infra, and it was very convenient to wrap the fastapi app thing with magnum to quickly get some of the advantages of lambda and reduce the drawbacks of ecs. I think that was a win. 
Going forward, I'm tring to understand whether we want to use server gateway interface behind a serverless infrastructure. I had a look at [this article](https://towardsdatascience.com/fastapi-aws-robust-api-part-1-f67ae47390f9) and the magnum docs and I'm still a bit confused. API gateway is the server gateway interface and lambdas are the route handlers. Vanilla lambdas are just as easy to write as any web framework route implementations. API gateway also lets us define [resource schemas](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/apigatewayv2_model) and probably does the boilerplate of schema validation. 
If I understand correctly that api gateway invokes a new lambda for each request, then we lose a lot of the benefit of as ASGI framework too, as it can't handle other requests while a given request awaits i/o.

Am I missing reasons _not_ to leverage api gateway for routing, resource schemas, access control, and rate limiting? Other than that it would be a moderately large change?

And then when, if ever, would it make sense to do this on apt/snwg? Maybe just something to consider for the future projects.

**Comment by leothomas** [comment]
@jo-tham This is super helpful, thanks! 

I don't have any cloudwatch metrics on hand - it'll require some spelunking through the Cloudwatch logs. I'd say the delay occurs infrequently, but is long enough to be truly frustrating. Here's a [slack thread](https://developmentseed.slack.com/archives/C04KHH66Q/p1630679321077800) that has a bit more info from when I first ran into this issue. 

With respect to the architecture remarks: 
> If I understand correctly that api gateway invokes a new lambda for each request, then we lose a lot of the benefit of as ASGI framework too, as it can't handle other requests while a given request awaits i/o.

100% agree. Another big thing lost from the ASGI framework is that background tasks are no longer executed in the background. Lambda functions stop executing once they return from the handler, so any operation that includes a background task has to first execute the background task and then return

That being said, for my understanding, leveraging API gateway for routing wouldn't restore the benefits of the ASGI framework, correct? 

The complexity I can forsee with using API Gateway for routing is that those routes will have to be defined both within the CDK deployment code  and within the application code itself, which may be error prone/messy to configure.

**Comment by leothomas** [comment]
Lastly, AWS provides a [Cognito integration for API Gateway](https://medium.com/@chandupriya93/providing-authorization-to-api-gateway-with-cognito-identity-pools-af451fd3b532) which completely removes all of the cognito boilerplate code to convert an authorization bearer token to a decoded JWT, _however_ this integration doesn't provide a way to authorize an unauthenticated user (ie: every request must have a auth token), which is a requirement in APT (since unauthenticated users can have access to the published ATBDs)

**Comment by jo-tham** [comment]
Thanks for the thoughts Leo 🙌 

> That being said, for my understanding, leveraging API gateway for routing wouldn't restore the benefits of the ASGI framework, correct?

That is my understanding, too. API GW doesn't give us an ASGI benefits. 

> The complexity I can forsee with using API Gateway for routing is that those routes will have to be defined both within the CDK deployment code and within the application code itself, which may be error prone/messy to configure.

I don't think this is the case. The routes would only be specified in the CDK code (I've only dont it with serverless&terraform, not cdk, I'm assuming cdk would be equivlent. The route definition refers to the lamba handler. The lambda handler, doesn't need to know what route refers to it. But it the developer does, i.e. implement the proper applicaiton logic within that handler, so there's an implicit knowledge of the routing in the route handlers in this framework. I admit, I haven't used it for APIs with very many resources, capturing this implicy routing knowledge may become really cumbersome.

> Lastly, AWS provides a [Cognito integration for API Gateway](https://medium.com/@chandupriya93/providing-authorization-to-api-gateway-with-cognito-identity-pools-af451fd3b532) which completely removes all of the cognito boilerplate code to convert an authorization bearer token to a decoded JWT, however this integration doesn't provide a way to authorize an unauthenticated user (ie: every request must have a auth token), which is a requirement in APT (since unauthenticated users can have access to the published ATBDs)

I'm not sure I totally understand this part. I think it could be done with API GW. Although it would be done at the route level, i.e. one route to list/read published atbds that doesn't require auth token. As opposed to having one route that filters atbds to only the published ones if the requester is anonymous. 

There's probably also the potential for a hybrid, where we use api gateway and cognito authentication and api gw decods the jwt and puts in in request headers, for us to lookup the cognito user's permissions.

I should also say it works damn well as is, and some other advantages already come to mind - I think definining the routing and schemas is faster with fastapi than with cdk/other iaac frameworks. Makes it easier to test locally, too. 

Though provoking!

**Comment by jo-tham** [comment]
@leothomas closer to the actual topic - in the thread you shared there was indication this delay is only on OPTIONS requests. 

Can you confirm whether that's the case?

If not, do you know exactly which other routes are affected?

**Comment by leothomas** [comment]
> one route to list/read published atbds that doesn't require auth token. As opposed to having one route that filters atbds to only the published ones if the requester is anonymous.

Yes! You're absolutely correct - this is probably the best solution to enable authenticated/unauthenticated access to documents. 

> so, it's not a latex/container issue?
The only reason I think it's a latex/container issue is because I've used the exact same setup: docker image lambda function with a Mangum wrapped FastAPI app, invoked through API Gateway and integrated with RDS/S3/etc, and none of those have similar delays when performing OPTIONS requests - the only difference being the addition of Latex libraries + compilers in docker image

**Comment by aboydnw** [comment]
Closing as not planned due to age and the problem being reduced via other means.
